qa/tasks/cephfs/test_scrub_checks.py

```diff
@@ -443,6 +444,50 @@ def test_scrub_split_dirfrags(self):
             timeout=30
         )
 
+    def test_scrub_backtrace_divergent_newer(self):
+        """
+        The consequence of the scrub will not acquire nest locks which would update
+        the iversion for the in-memory inode prior to comparing to on-disk values.
+        """
+
+        self.fs.set_max_mds(2)
+        self.fs.wait_for_daemons()
+
+        test_path2 = "testdir1/testdir2"
+        test_path3 = f"{test_path2}/testdir3"
+        test_path4 = f"{test_path3}/testdir4"
+        test_file = f"{test_path3}/file_0"
+
+        self.mount_a.run_shell(["mkdir", "-p", test_path4])
+        self.mount_a.setfattr(test_path2, "ceph.dir.pin", "0") # no surprise migrations
+
+        # Will trigger to load ancestor inodes in mds.1
+        self.mount_a.setfattr(test_path4, "ceph.dir.pin", "1")
+        self.mount_a.run_shell_payload(f"cd {test_path4} && for i in `seq 1 10`; do touch file_$i; done;")
+
+        # Dirty the ancestor inodes in mds.0 and then the mds.1 version will become older
+        self.mount_a.run_shell(["touch", f"{test_file}"])
+        self.mount_a.run_shell_payload(f"cd {test_path3} && for i in `seq 1 32`; do mkdir dir_$i; done;")
+
+        # Flush path 'testdir3/file_0' and this will persist the backtrace to disk
+        self.fs.rank_asok(["flush_path", f"/{test_file}"], rank=0)
```

Comment:
Can we not just flush rank 0's journal? I'm dubious about only flushing a specific path.

Reply:
No, this won't work. I only need to flush the `testdir3/file_0`, if we flush the whole journal it will also flush the ancestors'.

Reply:
Putting a comment I've been working on here for the last several days but at this point I don't want to lose it as I've been working on other things:

So this seems to be a key issue here. I'm not sure how in the original issue that `testdir3/file_0` would otherwise have its backtrace flushed. In general, journal trimming will unscatter the nestlock on parents as it commits to disk. That should flush the version to the replicas. I think we're missing a subtlety here.

This makes me think that the attractive option of "ignoring replicated metadata with out-of-date version" may actually be hiding a genuine bug.

Xiubo: I didn't get your test to run without some modification. I've since made numerous changes in an effort to simplify reproduction and forgot what needed adjustment in the first place.

Reply:
> Putting a comment I've been working on here for the last several days but at this point I don't want to lose it as I've been working on other things:
> 
> So this seems to be a key issue here. I'm not sure how in the original issue that `testdir3/file_0` would otherwise have its backtrace flushed. In general, journal trimming will unscatter the nestlock on parents as it commits to disk. That should flush the version to the replicas. I think we're missing a subtlety here.
> 
> This makes me think that the attractive option of "ignoring replicated metadata with out-of-date version" may actually be hiding a genuine bug.
> 
> Xiubo: I didn't get your test to run without some modification. I've since made numerous changes in an effort to simplify reproduction and forgot what needed adjustment in the first place.

How about when the journal event of `testdir3` and `file_0` are in different segments ?  And only the `file_0`'s events are flushed ?

Reply:
I don't think that's possible but if you can construct an example that may be it.

Reply:
If it's helpful, here's the test case I was actively playing with:

```
diff --git a/qa/tasks/cephfs/test_scrub_checks.py b/qa/tasks/cephfs/test_scrub_checks.py
index 15f3326f0ee..e4710011140 100644
--- a/qa/tasks/cephfs/test_scrub_checks.py
+++ b/qa/tasks/cephfs/test_scrub_checks.py
@@ -450,8 +450,67 @@ class TestScrubChecks(CephFSTestCase):
         the iversion for the in-memory inode prior to comparing to on-disk values.
         """
 
+        self.config_set('mds', 'mds_log_events_per_segment', 5)
+        self.config_set('mds', 'mds_log_max_segments', 8)
+
+        self.config_set('client', 'debug_client', 20)
+        self.config_set('client', 'debug_ms', 1)
+
         self.fs.set_max_mds(2)
-        self.fs.wait_for_daemons()
+        status = self.fs.wait_for_daemons()
+
+        if False:
+            self.mount_a.run_shell_payload("mkdir foo && mkdir foo/bar{0..32}")
+            # force fragmentation and distribution
+            self.mount_a.setfattr("foo", "ceph.dir.pin.distributed", "1")
+            self._wait_distributed_subtrees(4, status=status)
+
+            #self.fs.flush(rank=0)
+            self.mount_a.run_shell_payload("mkdir foo/bar{33..1024}")
+
+            time.sleep(20)
+
+            out_json = self.fs.run_scrub(["start", "/", "recursive"])
+            self.assertNotEqual(out_json, None)
+            self.wait_for_health("MDS_DAMAGE", 60)
+
+        #self.mount_a.run_shell_payload("mkdir -p foo/bar/baz && dd if=/dev/urandom of=foo/bar/baz/zab bs=1k count=1")
+        self.mount_a.run_shell_payload("mkdir -p foo/{bar0,bar1}/baz")
+        self.mount_a.setfattr("foo", "ceph.dir.pin", "0") # no surprise migrations
+        self.fs.flush(rank=0)
+        self.mount_a.setfattr("foo/bar1", "ceph.dir.pin", "1") # "foo" must be replicated on rank 1
+        #self._wait_subtrees([("/foo", 0), ("/foo/bar", 1)])
+        self._wait_subtrees([("/foo", 0), ("/foo/bar1", 1)])
+
+        import json
+        J1 = self.fs.rank_tell("dump", "tree", "foo", "0", rank=0)
+        log.info(f"{json.dumps(J1, indent=2)}")
+        J2 = self.fs.rank_tell("dump", "tree", "foo", "0", rank=1)
+        log.info(f"{json.dumps(J2, indent=2)}")
+        self.assertEqual(J1[0]['version'], J2[0]['version'])
+
+        #rados --pool cephfs.teuthology.data getxattr 10014bb9b3a.00000000 parent | ceph-dencoder type inode_backtrace_t import - decode dump_json
+
+        self.mount_a.run_shell_payload("cd foo/bar1/ && touch file && sync .")
+        self.mount_a.run_shell_payload("cd foo/bar0/ && for i in `seq 1 1000`; do touch file_$i; done;")
+        #self.fs.flush(rank=0)
+
+        J3 = self.fs.rank_tell("dump", "tree", "foo", "0", rank=0)
+        log.info(f"{json.dumps(J3, indent=2)}")
+        self.assertGreater(J3[0]['version'], J1[0]['version'])
+        J4 = self.fs.rank_tell("dump", "tree", "foo", "0", rank=1)
+        log.info(f"{json.dumps(J4, indent=2)}")
+        self.assertEqual(J4[0]['version'], J2[0]['version'])
+        self.assertGreater(J3[0]['version'], J4[0]['version'])
+
+        time.sleep(20)
+
+        out_json = self.fs.run_scrub(["start", "/", "recursive"])
+        self.assertNotEqual(out_json, None)
+        self.wait_for_health("MDS_DAMAGE", 60)
+
+        if True:
+            return
 
         test_path1 = "testdir1"
         test_path2 = "testdir1/testdir2"
@@ -462,6 +521,7 @@ class TestScrubChecks(CephFSTestCase):
         # Make sure all the ancestors' auth is in rank 0
         self.mount_a.run_shell(["mkdir", test_path1])
         self.mount_a.setfattr(test_path1, "ceph.dir.pin", "0") # no surprise migrations
+        subtrees = [(f"/{test_path1}", 0)]
         self.mount_a.run_shell(["mkdir", "-p", test_path4])
 
         # Will trigger to load ancestor inodes in mds.1
@@ -474,13 +534,16 @@ class TestScrubChecks(CephFSTestCase):
         self.mount_a.run_shell_payload(f"cd {test_path3} && for i in `seq 1 32`; do mkdir dir_$i; done;")
 
         # Flush path 'testdir3/file_0' and this will persist the backtrace to disk
-        self.fs.rank_asok(["flush_path", f"/{test_file}"], rank=0)
+        #self.fs.rank_asok(["flush_path", f"/{test_file}"], rank=0)
+        self.fs.flush(rank=0)
 
         # Export testdir3 to mds.1
-        self.mount_a.setfattr(test_path1, "ceph.dir.pin", "-1")
-        self.fs.rank_asok(["export", "dir", f"/{test_path3}", "1"], rank=0)
+        #self.mount_a.setfattr(test_path1, "ceph.dir.pin", "-1")
+        self.mount_a.setfattr(test_path3, "ceph.dir.pin", "1")
+        subtrees += [(f"/{test_path3}", 1)]
+        #self.fs.rank_asok(["export", "dir", f"/{test_path3}", "1"], rank=0)
         # wait for the export to settle
-        self._wait_subtrees([(f"/{test_path3}", 1)])
+        self._wait_subtrees(subtrees)
 
         # Scrub 'testdir3/file_0' and it will detect the backtrace in memory will be older and
         # metadata damage will be reported via cluster health warnings.
```

obviously it's a mess and I was testing various ideas. But maybe it'll inspire you.

Reply:
BTW, did this patch work ? Could you see the scrub error ?

Reply:
No, I don't see it because the nestlock is unscattered (gathered) during journal trimming. (I was playing with journal configs to get it to trim faster.) For that reason, I'm not really sure how this bug is reproduced and unfortunately we've garbage collected the original logs in the tracker ticket.

Reply:
Okay, maybe somewhere we should add some debug logs.

Reply:
`flush_path` is a dirty hack for a test to reproduce a problem arising from normal MDS operations. To my knowledge, it's not normally possible to flush an inode this way without also triggering a gather for parent inode's ifile/etc.

Again, we're missing something here that can only be explained with new debug logs with the original problem.

Reply:
> `flush_path` is a dirty hack for a test to reproduce a problem arising from normal MDS operations. To my knowledge, it's not normally possible to flush an inode this way without also triggering a gather for parent inode's ifile/etc.
> 
> Again, we're missing something here that can only be explained with new debug logs with the original problem.

The `flush_path` is simple:

```
13464 void MDCache::flush_dentry_work(const MDRequestRef& mdr)
13465 {
13466   MutationImpl::LockOpVec lov;
13467   CInode *in = mds->server->rdlock_path_pin_ref(mdr, true);                                                                                                                                                                                             
13468   if (!in)
13469     return;
13470  
13471   ceph_assert(in->is_auth());
13472   in->flush(new C_FinishIOMDR(mds, mdr));
13473 }
```
And in this case in `rdlock_path_pin_ref()` it won't acquire any ancestor inodes' locks unless there the `lock_cache` is set. And then no any chance to trigger the gathers for parent inodes.

Did I miss something about this ?